cran-skeleton: support optional licenses

dbast opened the pull request:
This is a bit tricky. Ideas welcome to improve it or do it in a different way.

Still I believe, that this gives some skeleton generated recipes a valid license file, which they
currently miss. This is not only helpful for new recipes, but also for existing ones, that get re-generated as part of an update process.

@jdblischak updated the license handling before, maybe you want to take a look.

This can be tested with e.g. r-dt or r-zoo.

Thanks

<!---
Thanks for opening a PR on conda-build!

Please include a news entry with your PR to help keep our changelog up to date!
There are instructions available at: https://regro.github.io/rever-docs/news.html

If there is specific issue / feature request that this PR is addressing,
please link to the corresponding issue by using the `#issuenumber` syntax.

Thanks again!
-->

jdblischak commented:
@dbast Thanks for working on this! I've been trying to think about how to best improve the coverage of automatic adding license files to recipes, and I agree that catching the `|` cases is a good option.

Could you please add tests for `r-dt` and `r-zoo`? You can add them to the list below:

https://github.com/conda/conda-build/blob/63643da4ecc5c5b9bca0433eb9dcd3cf7195dfb5/tests/test_api_skeleton.py#L198-L204

jdblischak commented on conda_build/skeletons/cran.py:
```diff
@@ -1423,13 +1423,20 @@ def get_license_file(license_text):
+      * Single unambiguous complete license: 'GPL-2' -> 'GPL-2'
+      * Optional complete and custom license: 'GPL-2 | file LICENSE' -> 'GPL-2'
+      * Two optional complete licenses: 'GPL-2 | GPL-3' -> 'GPL-2'
+      * Two optional complete licenses: 'GPL-3 | GPL-2' -> 'GPL-2'
```
Is favoring `GPL-2` in this final case a technical or legal reason? In other words, why is this behavior chosen over these two alternatives:

1. Always choose the first license that the package author lists
1. Always favor GPL-3 over GPL-2 since it is the latest version of the GPL

dbast replied:
I thought there are two options for the match order:
1. The order of the licenses in `license_text`: In this case the package upstream can influence it.
2. The order of the sorted keys of `d_license`: In this case the cran skeleton defines the order (which could be also legal driven)

I just did 2, because thats what I just started with :) And I also like that both `GPL-2 | GPL-3` (now new) and `GPL (>= 2)` (was already the case before) result in GPL2, even though both could match to GPL3... just for a kind of consistency... not because I prefer the one over the other... 

I am also happy to change the implementation to approach 1. And of course adding tests makes sense.

Most important thing is that the matching is deterministic and resulting in a correct license.

(Best thing would be maybe to include multiple license files, that would also allow to add combined licenses and the match oder would not matter anymore, but that requires probably a change of the meta.yaml schema to allow `license_file` being a list.. sounds like a bigger exercise)

jdblischak replied:
>    The order of the licenses in license_text: In this case the package upstream can influence it.

I don't follow. How can the upstream package influence it? If the authors put a given license as the first choice, I think it is safe to assume that we can use it.

>    The order of the sorted keys of d_license: In this case the cran skeleton defines the order (which could be also legal driven)

This seems arbitrary since it is alphabetical.

> I am also happy to change the implementation to approach 1. And of course adding tests makes sense.

To test `GPL-3 | GPL-2`, you can use `r-magree`. It's actually the only package I could find with that exact license.

> Most important thing is that the matching is deterministic and resulting in a correct license.

Agree 100%. The current problem is that my `for` loop solution worked well for the single license case, but isn't very sophisticated. To handle multiple licenses, it may be necessary to rethink the approach.

> (Best thing would be maybe to include multiple license files, that would also allow to add combined licenses and the match order would not matter anymore, but that requires probably a change of the meta.yaml schema to allow license_file being a list.. sounds like a bigger exercise)

Agreed. I had investigated this possibility. Since it is only possible to include one file, I was thinking long-term we could concatenate files when necessary (e.g. for the common case of the MIT license, concatenate the file `LICENSE` on top of the MIT template file). But since that would require reading and writing from disk, I didn't attempt that in my initial PR.

dbast replied:
ok. Found some time to change the matching to respect the given license and extended the tests.

dbast replied:
Closing as this was now done in a different way.

dbast commented:
The CI failures seem to be un-related. Ready for review/merge ... thanks :)

Both commits can be squashed into one, they show different implementations. Can also do the squash on my side.

jdblischak commented on tests/test_api_skeleton.py:
```diff
@@ -201,6 +201,9 @@ def test_pypi_section_order_preserved(testing_workdir):
                  ('r-cortools', 'Artistic License 2.0', 'OTHER', 'Artistic-2.0'),
                  ('r-udpipe', 'MPL-2.0', 'OTHER', ''),
+                 ('r-dt', 'GPL-3 | file LICENSE', 'GPL3', 'GPL-3'),
+                 ('r-zoo', 'GPL-2 | GPL-3', 'GPL3', 'GPL-2'),
```
This is unfortunate. The license family should be `GPL2`. I think the problem is that the full license text `GPL-2 | GPL-3` gets passed to `guess_license_family` ([source](https://github.com/conda/conda-build/blob/master/conda_build/skeletons/cran.py#L1100)), and then `guess_license_family` searches for `GPL-3` prior to searching for `GPL-2` ([source](https://github.com/conda/conda-build/blob/master/conda_build/license_family.py#L85))

dbast replied:
Right. This behaviour was already there before. Let's see what I can do about it.

dbast replied:
Ok. Somebody invested some time in `guess_license_family` to come up with a specific matching order. Will just follow that.

jdblischak replied:
@dbast Could you share links to the previous discussion on the license order? I'm curious why this order was chosen.

dbast replied:
@jdblischak Well, I just read https://github.com/conda/conda-build/blame/master/conda_build/license_family.py and saw that several people worked on it.... and just assumed that there is some reasoning behind.... maybe there some extensiv discussion in some of the merged PRs...  Also the order is not obvious.... it is first trying to match GPL3, then GPL2 (like newest to oldest) and if that is not successful, following the order of `allowed_license_families` in https://github.com/conda/conda-build/blame/master/conda_build/license_family.py#L8

dbast replied:
Other option would be calling `guess_license_family` based on the matching of `get_license_file` ...  That would be a normalization as suggested in https://github.com/conda/conda-build/blob/master/conda_build/skeletons/cran.py#L1098

jdblischak replied:
@dbast OK, thanks for the explanation. I think your decision to follow the current convention for the skeletons makes sense.

dbast replied:
Closing as this was now done in a different way.

dbast commented:
Update:
* First commit now introduces more tests first to show current behavior before changing it.
* Second commit implements matching of optional licenses in the order as done by `guess_license_family`

jdblischak commented:
> matching of optional licenses in the order as done by guess_license_family

@dbast That decision works for me!

jdblischak commented:
@msarahan @katietz @mingwandroid Could you please review this PR to improve the license file capability of the R/CRAN skeleton?

jdblischak commented:
@dbast I found an example where the current version of the CRAN skeleton has the same issue with the mismatched `license_family` and `license_file`. The R package [MGLM](https://cran.r-project.org/package=MGLM) has the license string `GPL (>= 2)`. In this case, `guess_license_family` returns `GPL3` but `get_license_file` returns `GPL-2`.

dbast commented:
Completely reworked this and rebased it on top of the nice work of @isuruf https://github.com/conda/conda-build/pull/3616

dbast commented:
cc @bgruening, @cbrueffer, @daler, @dbast, @dpryan79, @jdblischak, @johanneskoester

dbast commented:
Ready for review.

dbast commented:
Some examples:
```
about:
  home: http://github.com/tidyverse/broom
  license: MIT
  summary: Summarizes ...
  license_family: MIT
  license_file:
    - '{{ environ["PREFIX"] }}/lib/R/share/licenses/MIT'
    - LICENSE
```

```
about:
  home: https://CRAN.R-project.org/package=corTools
  license: Artistic-2.0
  summary: Designed ...
  license_family: OTHER
  license_file:
    - '{{ environ["PREFIX"] }}/lib/R/share/licenses/Artistic-2.0'
```

```
about:
  home: https://CRAN.R-project.org/package=magree
  license: GPL-3 | GPL-2
  summary: Implements ...
  license_family: GPL3
  license_file:
    - '{{ environ["PREFIX"] }}/lib/R/share/licenses/GPL-3'
    - '{{ environ["PREFIX"] }}/lib/R/share/licenses/GPL-2'
``` 

```
about:
  home: https://bnosac.github.io/udpipe/en/index.html, https://github.com/bnosac/udpipe
  license: MPL-2.0
  summary: This ...
  license_family: OTHER
```

dbast commented:
@jdblischak The r-mglm example:

```
about:
  home: https://CRAN.R-project.org/package=MGLM
  license: GPL-2
  summary: Provides ...
  license_family: GPL2
  license_file:
    - '{{ environ["PREFIX"] }}/lib/R/share/licenses/GPL-2'
```

jdblischak commented:
r-mglm looks good. Could you please also add it as a test case since we don't currently have one for `GPL (>= 2)`?

What about r-dt? You mentioned it earlier since it has a [second license file](https://cran.r-project.org/web/packages/DT/LICENSE). When I tested your branch locally, it only added the GPL-3 license. Shouldn't it instead be:

```
about:
  home: https://rstudio.github.io/DT
  license: GPL-3
  summary: Data objects in R can be rendered as HTML tables using the JavaScript library 'DataTables'
    (typically via R Markdown or Shiny). The 'DataTables' library has been included
    in this R package. The package name 'DT' is an abbreviation of 'DataTables'.
  license_family: GPL3
  license_file:
    - '{{ environ["PREFIX"] }}/lib/R/share/licenses/GPL-3'
    - LICENSE
```

Regardless of the final decision, could you please add r-dt as an additional test case? That way the expected behavior in this case is documented.

dbast commented:
@jdblischak Thanks for testing and your feedback.

Updates:
* Added testcase for r-mglm: Behaviour is not changed by this PR.
* Added testcase for r-dt with explanation of "XXX + file LICENSE"  vs. "XXX | file LICENSE" handling: 
  ```
  It is expected that the file "LICENSE" is explicitly not added to
  license_file, as it is not extending a to be combined license (that would
  be "XXX + file", instead of "XXX | file") and the content of the file is
  unknown in regards to the name/id or family.
  ```

dbast commented:
Test should be now more clear in regards to what they match to what by including the cran meta data.

isuruf commented:
> It is expected that the file "LICENSE" is explicitly not added to
license_file, as it is not extending a to be combined license

This is wrong. `LICENSE` should be added to the `license_file`. `r-dt` is a GPL package, but it uses packages like jquery.highlight.js which is under MIT and with MIT the copyright notice in `LICENSE` must be distributed.

jdblischak commented:
I tend to agree with @isuruf that we should aim to include all the license files that the authors describe. This will likely lead to some false positives. For example, I've seen cases where authors put "GPL-3 | LICENSE", where LICENSE is simply the GPL-3 text. But the only downside there is a slightly larger tarball. In more complex case like r-dt, I think it makes sense to include the additional license for the vendored software, even if it isn't reflected in license_family.

dbast commented:
Ok. I agree that the LICENSE file should be added in case of r-dt.

Though the idea oft this PR was to split "XXX | YYY" by "|" and then parse each part with the same logic as before. 

If we agree, that we should add as many license files as possible to detect to be on the safe side, I am happy to update also the parsing logic to support "file XXX" as in r-dt.

(The opposite would be to add no license file at all, if some parts can not be clearly matched, to clearly signal that manual work is required. Which is no improvement, if that manual work is just then not done.)

marcelotrevisani commented:
I think a similar work might be done for the ``pypi`` skeleton. But it can be done in a second step.
There is another license which we should think about it which is the ``ISC license``.

That is described on #3645 
But it also can be done after it.

cc: @msarahan 

I am going to merge it then, thanks for the effort! :)

marcelotrevisani commented:
I just realized that, there is a small linter problem regarding flake8
Could you please fix it? After that it should be good to go! :)

dbast commented:
Updated to support r-dt case, which is now:
```
about:
  home: https://rstudio.github.io/DT
  license: GPL-3
  summary: Data...
  license_family: GPL3
  license_file:
    - '{{ environ["PREFIX"] }}/lib/R/share/licenses/GPL-3'
    - LICENSE
```

Also fixed the flake8 findings.

dbast commented:
The `ISC license` is not supported by cran according to https://svn.r-project.org/R/trunk/share/licenses/license.db ... so no urgent need to to add it here.

dbast commented:
Updated once again. Good to go from my side.

marcelotrevisani commented:
There is a problem on our CI that ``six`` is not being installed, but it is not related to this PR
So, I am going to merge it then.

Thanks!
